src/core/vanilla.ts

```diff
@@ -396,129 +396,80 @@ const invalidateDependents = <Value>(state: State, atom: Atom<Value>): void => {
   atom: WritableAtom<Value, Update>,
-  update: Update,
-  pendingPromises: Promise<void>[]
+  update: Update
 ): void => {
-  const isPendingPromisesExpired = !pendingPromises.length
   const writePromise = getAtomState(state, atom)?.w
   if (writePromise) {
-    const promise = writePromise.then(() => {
-      writeAtomState(state, atom, update, pendingPromises)
-      if (isPendingPromisesExpired) {
-        flushPending(state)
-      }
+    writePromise.then(() => {
+      writeAtomState(state, atom, update)
+      flushPending(state)
     })
-    if (!isPendingPromisesExpired) {
-      pendingPromises.push(promise)
-    }
     return
   }
-  try {
-    const promiseOrVoid = atom.write(
-      ((a: AnyAtom) => {
-        const aState = readAtomState(state, a)
-        if (aState.e) {
-          throw aState.e // read error
-        }
-        if (aState.p) {
-          if (
-            typeof process === 'object' &&
-            process.env.NODE_ENV !== 'production'
-          ) {
-            console.warn(
-              'Reading pending atom state in write operation. We throw a promise for now.',
-              a
-            )
-          }
-          throw aState.p // read promise
-        }
-        if ('v' in aState) {
-          return aState.v // value
-        }
+  const promiseOrVoid = atom.write(
+    ((a: AnyAtom) => {
+      const aState = readAtomState(state, a)
+      if (aState.e) {
+        throw aState.e // read error
+      }
+      if (aState.p) {
         if (
           typeof process === 'object' &&
           process.env.NODE_ENV !== 'production'
         ) {
           console.warn(
-            '[Bug] no value found while reading atom in write operation. This probably a bug.',
+            'Reading pending atom state in write operation. We throw a promise for now.',
             a
           )
         }
-        throw new Error('no value found')
-      }) as Getter,
-      ((a: AnyWritableAtom, v: unknown) => {
-        const isPendingPromisesExpired = !pendingPromises.length
-        if (a === atom) {
-          if (!hasInitialValue(a)) {
-            // NOTE technically possible but restricted as it may cause bugs
-            throw new Error('no atom init')
-          }
-          setAtomValue(state, a, v)
-          invalidateDependents(state, a)
-        } else {
-          writeAtomState(state, a, v, pendingPromises)
-        }
-        if (isPendingPromisesExpired) {
-          flushPending(state)
-        }
-      }) as Setter,
-      update
-    )
-    if (promiseOrVoid instanceof Promise) {
-      const promise = promiseOrVoid.finally(() => {
-        setAtomWritePromise(state, atom)
-        if (isPendingPromisesExpired) {
-          flushPending(state)
+        throw aState.p // read promise
+      }
+      if ('v' in aState) {
+        return aState.v // value
+      }
+      if (
+        typeof process === 'object' &&
+        process.env.NODE_ENV !== 'production'
+      ) {
+        console.warn(
```

Comment:
Typo: this is probably a bug.

Reply:
oops. thanks. it has been a typo for a long time.